File: sdk/go/nhtml.go
package nhtml

import (
	"encoding/json"
	"io"
	"net/http"
)

const Version = "0.7.3-stable"

type Event struct {
	Handler     string                 `json:"handler"`
	SourceID    string                 `json:"source_id"`
	SessionID   string                 `json:"session_id"`
	Payload     string                 `json:"payload"`
	LastVersion uint32                 `json:"last_version"`
	Nodes       map[string]interface{} `json:"nodes"`
	Data        map[string]interface{} `json:"-"`
}

// Nhtml is the main entry point for the Go SDK
type Nhtml struct{}

func (n Nhtml) Patch() *Patch {
	return NewPatch()
}

func (n Nhtml) ParseEvent(r *http.Request) (*Event, error) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		return nil, err
	}

	var event Event
	if err := json.Unmarshal(body, &event); err != nil {
		return nil, err
	}

	if event.Payload != "" {
		var payloadData map[string]interface{}
		if err := json.Unmarshal([]byte(event.Payload), &payloadData); err == nil {
			event.Data = payloadData
		}
	}

	return &event, nil
}

func (n Nhtml) Send(w http.ResponseWriter, p *Patch) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(p)
}

func (n Nhtml) JoinRoom(w http.ResponseWriter, roomID string) {
	n.Send(w, NewPatch().JoinRoom(roomID))
}

func (n Nhtml) LeaveRoom(w http.ResponseWriter, roomID string) {
	n.Send(w, NewPatch().LeaveRoom(roomID))
}

File: sdk/python/nhtml/nhtml.py
import json
from typing import List, Dict, Any, Optional, Callable
from .patch import Patch

class Nhtml:
    @staticmethod
    def patch() -> Patch:
        """Creates a new patch collection."""
        return Patch.create()

    @staticmethod
    def batch(callback: Callable[[Patch], None]) -> Dict[str, Any]:
        """Starts a batch operation, executes the callback, and returns the dict response."""
        p = Patch.create()
        callback(p)
        return p.to_dict()

    @staticmethod
    def parse_event(body: Dict[str, Any]) -> Dict[str, Any]:
        """Parses a raw NHTML event from the request body."""
        payload = body.get('payload', '')
        data = {}
        try:
            if payload:
                data = json.loads(payload)
        except json.JSONDecodeError:
            pass
            
        return {
            'handler': body.get('handler'),
            'source_id': body.get('source_id'),
            'session_id': body.get('session_id'),
            'payload': payload,
            'data': data,
            'last_version': body.get('last_version', 0),
            'nodes': body.get('nodes', {})
        }

    @staticmethod
    def join_room(room_id: str) -> Dict[str, Any]:
        """Utility to join a room immediately."""
        return Patch.create().join_room(room_id).to_dict()

    @staticmethod
    def leave_room(room_id: str) -> Dict[str, Any]:
        """Utility to leave a room immediately."""
        return Patch.create().leave_room(room_id).to_dict()
